### PR TITLE
Add MDC entry writers to be able to convert values

### PR DESCRIPTION
**Pull request by jug**
Solution proposal for issue https://github.com/logfellow/logstash-logback-encoder/issues/955

### CODE REVIEW

**Review by philsttr — CHANGES_REQUESTED** [review]
Thanks for the contribution!   I really like the approach you came up with.  I have some feedback on naming and some low level implementation details.  But in general, the design is solid.

File: src/main/java/net/logstash/logback/composite/loggingevent/mdc/BooleanMdcFieldConverter.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+    @Override
+    public boolean writeMdcField(JsonGenerator generator, MdcField mdcField) throws IOException {
+        if (PATTERN_BOOLEAN.matcher(mdcField.getValue()).matches()) {
+            generator.writeFieldName(mdcField.getAlternateName());
+            generator.writeBoolean(Boolean.parseBoolean(mdcField.getValue()));
+            return true;
+        }
+
+        return false;
```
[the remark this answers is not in the document]
Comment (jug) [reply src/main/java/net/logstash/logback/composite/loggingevent/mdc/BooleanMdcFieldConverter.java:35]: yes, that works too (and is simpler).

File: src/main/java/net/logstash/logback/composite/loggingevent/mdc/LongMdcFieldConverter.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    public boolean writeMdcField(JsonGenerator generator, MdcField mdcField) throws IOException {
+        try {
+            long parsedValue = Long.parseLong(mdcField.getValue());
+            generator.writeFieldName(mdcField.getAlternateName());
+            generator.writeNumber(parsedValue);
+            return true;
+        } catch (NumberFormatException ignore) {
+        }
+
+        return false;
```
[the remark this answers is not in the document]
Comment (jug) [reply src/main/java/net/logstash/logback/composite/loggingevent/mdc/LongMdcFieldConverter.java:34]: I've originally used the parsing right away as I don't need to know what makes a Long or Double and the try-catch is needed regardless ... a number greater Long.MAX_VALUE looks like a number, but cannot be parsed into a Long.

But I agree that lots of Exceptions are catched.
Will add a preliminary check.

File: src/main/java/net/logstash/logback/composite/loggingevent/mdc/MdcField.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+ */
+package net.logstash.logback.composite.loggingevent.mdc;
+
+public class MdcField {
+
+    private final String name;
+    private final String alternateName;
+    private final String value;
+
+    public MdcField(String name, String alternateName, String value) {
```
[the remark this answers is not in the document]
Comment (jug) [reply src/main/java/net/logstash/logback/composite/loggingevent/mdc/MdcField.java:24]: I was also not keen on the name `alternateName`, but didn't quite wrap my head around for something more speaking.
 
will do

File: src/main/java/net/logstash/logback/composite/loggingevent/mdc/DoubleMdcFieldConverter.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+    @Override
+    public boolean writeMdcField(JsonGenerator generator, MdcField mdcField) throws IOException {
+        try {
+            double parsedValue = Double.parseDouble(mdcField.getValue());
+            generator.writeFieldName(mdcField.getAlternateName());
+            generator.writeNumber(parsedValue);
+            return true;
+        } catch (NumberFormatException ignore) {
+        }
```
[the remark this answers is not in the document]
Comment (jug) [reply src/main/java/net/logstash/logback/composite/loggingevent/mdc/DoubleMdcFieldConverter.java:32]: ok ... will add check

File: src/main/java/net/logstash/logback/composite/loggingevent/mdc/MdcFieldConverter.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package net.logstash.logback.composite.loggingevent.mdc;
+
+import java.io.IOException;
+
+import com.fasterxml.jackson.core.JsonGenerator;
+
+public interface MdcFieldConverter {
```
[the remark this answers is not in the document]
Comment (jug) [reply src/main/java/net/logstash/logback/composite/loggingevent/mdc/MdcFieldConverter.java:22]: makes sense ... will rename

File: src/main/java/net/logstash/logback/composite/loggingevent/mdc/MdcField.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package net.logstash.logback.composite.loggingevent.mdc;
+
+public class MdcField {
```
[the remark this answers is not in the document]
Comment (jug) [reply src/main/java/net/logstash/logback/composite/loggingevent/mdc/MdcField.java:18]: make sense ... will remove

File: src/main/java/net/logstash/logback/composite/loggingevent/mdc/MdcFieldConverter.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+    /**
+     * Writes the given {@link MdcField} allowing to manipulate the output of the field name and field value.
+     *
+     * @param mdcField all attributes of {@link MdcField} are non-null.
+     *
+     * @return Return true if the converter handled the output of the field, otherwise return false.
+     */
+    boolean writeMdcField(JsonGenerator generator, MdcField mdcField) throws IOException;
```
[the remark this answers is not in the document]
Comment (jug) [reply src/main/java/net/logstash/logback/composite/loggingevent/mdc/MdcFieldConverter.java:31]: come by on refactoring for the other changes anyway

**Review by philsttr — APPROVED** [review]

**Comment by philsttr** [comment]
Thanks for the quick turnaround on my review comments!  This will be included in the next release (soon)

**Pull request merged**